[Automation-Tier2] Verify Deletion of Protected snapshot on RBD image

preethisn opened the pull request:
Jira ticket -https://issues.redhat.com/browse/RHCEPHQE-7783
Test case - On krbd client - Performs deletion of protected snapshot on a RBD image
https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-9224

Test flow:
1) Create a pool and an Image 
2) using KRBD client, map the image and run IOs using FIO 
3) Perform snapshot create on a RBD image
3) Perform snapshot protect for the image
4) Perform clone of af an image, Have mulitple clones 
5) Perform Delete of a snapshot which is protected 
6) Cover the above steps on EC pool as well
Expected result-
at step 6 - Delete of a snap should fail since it is protected

Test Results:
5.3 - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-7OGN7G
6.0- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-NST9O7

sunilangadi2 commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,95 @@
+            else:
+                log.info(f"Clone {clone_name} created successfully")
+
+        if rbd.snap_remove(pool, image, snap):
```
what if bug in code and snapshot gets deleted?
Can you please handle that with a proper error message and make test to fail?

sunilangadi2 replied:
@preethisn `error` and `info` message needs to reverse as it's negative scenario.

sunilangadi2 commented on suites/pacific/rbd/tier-2_rbd_regression.yaml:
```diff
@@ -161,3 +161,17 @@ tests:
+          io: true
+          nounmap: true
+      module: rbd_delete_protected_snapshot_krbd.py
+      name: krbd client - Perform Deletion of protected snapshot
```
Please start test names with  keyword `Test`  e.g `Test to perform Deletion of protected snapshot`

rkhambad commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+        kw.get("config")["image_spec"] = [image_spec]
+        kw.get("config")["size"] = size
+        kw["rbd_obj"] = rbd
+        krbd_io_handler(**kw)
```
krbd_io_handler will again try to create a image in specified image_spec , we have provision to pass "do_not_create_image: True" in initial_rbd_config method, we can make use of that and avoid creating image while creating pool.

preethisn replied:
Added the changed in the yaml under config.

sunilangadi2 commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+        log.info("Executing test on Replication pool")
+        if rbd_delete_protect_snap(rbd_obj.get("rbd_reppool"), "rep_pool_config", **kw):
+            return 1
+        log.info("Executing test on EC pool")
```
if EC pool creation fails then it will fail in `initial_rbd_config` itself right,also most of the earlier modules uses this approach only.

preethisn replied:
yes, am not changing this as i feel this approach is better. we need to fail the test if ec pool fails

preethisn commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+        if rbd.snap_remove(pool, image, snap):
+            log.error(f"Removal of protected {snap} failed")
+            return 0
+        log.info(f"Successfully removed {snap}")
```
done

preethisn commented:
Addressed  the review commnets.
Attched latest logs after addressesing the 2nd review - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-GHPLHM/krbd_client_-_Test_to_Perform_Deletion_of_protected_snapshot_0.log

rkhambad commented on suites/pacific/rbd/tier-2_rbd_regression.yaml:
```diff
@@ -161,3 +161,18 @@ tests:
+      module: rbd_delete_protected_snapshot_krbd.py
+      name: krbd client - Test to Perform Deletion of protected snapshot
+      polarion-id: CEPH-9224
+
```
one small change , extra lines at the end of the file.
Other changes look good

preethisn replied:
done

rkhambad reviewed (dismissed):
changes look good to me

sunilangadi2 reviewed (dismissed):
LGTM

VasishtaShastry commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+                log.error(f"Creation of clone {clone_name} failed")
+                return 1
+            else:
+                log.info(f"Clone {clone_name} created successfully")
```
[optional]

IMO, this info of clone creation can be pushed to debug or removed.

preethisn replied:
done

VasishtaShastry commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+                log.info(f"Clone {clone_name} created successfully")
+
+        if rbd.snap_remove(pool, image, snap):
+            log.error(f"Removal of protected {snap} failed")
+            return 0
+        log.info(f"{snap} snapshot removed even after the snapshot is protected")
+        return 1
```
Since we have a negative TC here, we need to reverse the info and error here

Failure of snap removal is an info and being able to delete image is an ERROR from this TC perspective here, otherwise report portal might get confused along with others looking at the logs.

VasishtaShastry commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+    CEPH-9224 - Trying to delete a protected snapshot will fail.
+    Pre-requisites :
+    1. Cluster must be up and running with capacity to create pool
+       (At least with 64 pgs)
```
Can we please remove this line ? I think its redundant.

sunilkumarn417 commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+
+def rbd_delete_protect_snap(rbd, pool_type, **kw):
+    """
+    This module on krbd client - Performs deletion of protected snapshot on a RBD image
```
please update the doc string , Since this is one of the step in validation.

preethisn replied:
done

sunilkumarn417 commented on suites/quincy/rbd/tier-2_rbd_regression.yaml:
```diff
@@ -160,3 +160,15 @@ tests:
       polarion-id: CEPH-9836
+
+  - test:
+      desc: Trying to delete a protected snapshot should fail
```
please add Negative test in the description

preethisn replied:
done

sunilkumarn417 commented on tests/rbd/rbd_delete_protected_snapshot_krbd.py:
```diff
@@ -0,0 +1,97 @@
+            else:
+                log.info(f"Clone {clone_name} created successfully")
+
+        if rbd.snap_remove(pool, image, snap):
```
please add stderror validation

preethisn replied:
done

preethisn commented:
latest logs attached after addressing the review comments - All test logs located here: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1PFRY4

sunilkumarn417 commented:
please clear the conflicts

preethisn commented:
Updated the latest logs after resolving the conflicts-
[protected_snapshot.log](https://github.com/red-hat-storage/cephci/files/11048503/protected_snapshot.log)

Manohar-Murthy reviewed (approved):
team review is done.

sunilkumarn417 commented on tests/rbd/rbd_utils.py:
```diff
@@ -306,7 +307,14 @@ def snap_remove(self, pool_name, image_name, snap_name, **kw):
             cmd = f"rbd snap rm {pool_name}/{image_name}@{snap_name}"
-        return self.exec_cmd(cmd=cmd)
+        out, err = self.exec_cmd(cmd=cmd, all=True, check_ec=False)
+        if "Removing snap: 0% complete...failed" not in err:
```
please revert back to calling module, this would affect other test cases.
